src/noit_message_decoder.c

```diff
@@ -0,0 +1,165 @@
+    metric_value_t *metric, mtev_boolean has_noit) {
+  const char *cp, *metric_type_str, *time_str, *check_id_str;
+  char *value_str;
+  char *dp, id_str_copy[UUID_STR_LEN + 1];
+  cp = payload;
+
+  // Go to the timestamp column
+  MOVE_TO_NEXT_TAB(cp, time_str);
+  if(has_noit == mtev_true) { // non bundled messages store the source IP in the second column
+    if(time_str == NULL)
+      return -1;
+    MOVE_TO_NEXT_TAB(cp, time_str);
+  }
+
+  if(time_str == NULL)
+    return -1;
+  /* extract time */
+  metric->whence_ms = strtoull(time_str, &dp, 10);
+  metric->whence_ms *= 1000; /* s -> ms */
+  if(dp && *dp == '.')
+    metric->whence_ms += (int) (1000.0 * atof(dp));
+
+  MOVE_TO_NEXT_TAB(cp, check_id_str);
+  if(!check_id_str)
+    return -2;
+  MOVE_TO_NEXT_TAB(cp, *metric_name);
+  if(!*metric_name)
+    return -3;
+
+  memcpy(id_str_copy, *metric_name - UUID_STR_LEN - 1, UUID_STR_LEN);
+  id_str_copy[UUID_STR_LEN] = '\0';
+
+  if(uuid_parse(id_str_copy, *id) != 0) {
+    mtevL(mtev_error, "uuid_parse(%s) failed\n", id_str_copy);
+    return -7;
+  }
+
+  if(*metric_name - check_id_str < UUID_STR_LEN + 1)
+    return -6;
+
+  if(*payload == 'M') {
+    MOVE_TO_NEXT_TAB(cp, metric_type_str);
+    if(!metric_type_str)
+      return -4;
+    MOVE_TO_NEXT_TAB(cp, value_str);
+    if(!value_str)
+      return -5;
+
+    *metric_name_len = metric_type_str - *metric_name - 1;
+
+    switch (*metric_type_str) {
+    case METRIC_INT32:
+      metric->type = METRIC_INT32;
+      metric->value.v_int32 = strtol(value_str, NULL, 10);
+      break;
+    case METRIC_UINT32:
+      metric->type = METRIC_UINT32;
+      metric->value.v_uint32 = strtoul(value_str, NULL, 10);
+      break;
+    case METRIC_INT64:
+      metric->type = METRIC_INT64;
+      metric->value.v_int64 = strtoll(value_str, NULL, 10);
+      break;
+    case METRIC_UINT64:
+      metric->type = METRIC_UINT64;
+      metric->value.v_uint64 = strtoull(value_str, NULL, 10);
+      break;
+    case METRIC_DOUBLE:
+      metric->type = METRIC_DOUBLE;
+      metric->value.v_double = strtod(value_str, NULL);
+      break;
+    case METRIC_STRING:
+      metric->type = METRIC_STRING;
+      if(!strcmp(value_str, "[[null]]"))
+        metric->value.v_string = NULL;
+      else {
+        metric->value.v_string = value_str;
+      }
+      break;
+    default:
+      return -9;
+    }
+    return 1;
+  }
+
+  if(*payload == 'H') {
+    MOVE_TO_NEXT_TAB(cp, value_str);
+    if(!value_str)
+      return -4;
+
+    *metric_name_len = value_str - *metric_name - 1;
+
+    int vstrlen = strlen(value_str);
+
+    while ((vstrlen > 1) && (value_str[vstrlen - 1] == '\n'))
+      vstrlen--;
+    metric->type = METRIC_STRING;
+    if(vstrlen == 0)
+      metric->value.v_string = NULL;
+    else {
+      metric->value.v_string = value_str;
+    }
+    return 1;
+  }
+
+  if(*payload == 'S' || *payload == 'C' || *payload == 'D') {
+    metric->type = METRIC_STRING;
```

Comment:
The S, C and D records aren't metrics at all... this is a sloppy abstraction.

the problem here is that the directory is really a message director I think trying to squeeze non-metric types back into a strongly typed metric is a bad idea.